Fix several Serializer and Deserializer bugs, especially const

david-eckardt-sociomantic opened the pull request:
Fix #79 and a few more bugs I found on the way, also clean up the code a bit.
I made it so that the elements of a `const(T[])[]` array (which are of type `const(T[])`) will reside in the slices buffer which by itself is _not_ `const` but should never directly be written to anyway.

david-eckardt-sociomantic committed:
Serializer: Fix compile error for T[][n][]

gavin-norman-sociomantic commented:
If this is a bug fix, it should be based on v2.6.x.

gavin-norman-sociomantic commented:
>Use `ref` iteration over struct fields

Could you add an explanation of why this change is made, in the commit message?

gavin-norman-sociomantic commented on src/ocean/util/serialize/contiguous/Serializer.d:
```diff
@@ -290,7 +290,7 @@ struct Serializer
                         // Recurse into static array elements which contain a
                         // dynamic array.
 
-                        foreach (element; s.tupleof[i])
```
This change (and other similar ones, below) looks unrelated to "Use `ref` iteration over struct fields".

david-eckardt-sociomantic replied:
This commit replaces
```d
foreach (i, T; typeof(s.tupleof)) { /* use s.tupleof[i] to access the i-th field of s */ }
```
with
```d
foreach (i, ref field; s.tupleof) { /* use `field` to access the i-th field of s */ }
```
So it changes both the `foreach` iteration variable to `ref field` and `s.tupleof[i]` to `field` in the loop body.
This particular line here adjusts `s.tupleof[i]` -> `field` that seems to have been missed when the same change had been applied in a previous commit.
Or did you spot a change in this commit that does something different?

gavin-norman-sociomantic replied:
Ah I see. I didn't get the connection between the use of `ref` and the change of `s.tupleof[i]` -> `field`. Maybe just add a bit of explanation in the commit message.

david-eckardt-sociomantic replied:
Ironically this loop has always been broken in two ways: Doing the `foreach` loop is wrong, and in the loop `element` should be passed, not `field`.

gavin-norman-sociomantic commented:
>Deserializer: Fix deserialising qualified (i.e. const) arrays
>Serializer: Fix qualified dynamic arrays not detected

Would be nice to have unittests to confirm these fixes.

gavin-norman-sociomantic commented:
>Unqual whole array at once rather than each element at a time

Could you add an explanation of why this change is made, in the commit message?

david-eckardt-sociomantic commented:
> Would be nice to have unittests to confirm these fixes.

I added it to the `ocean.util.serialize.contiguous.package_` test.

gavin-norman-sociomantic commented:
>I added it to the ocean.util.serialize.contiguous.package_ test.

In the same commits?

david-eckardt-sociomantic commented:
> In the same commits?

Ha, no. One commit too late.

gavin-norman-sociomantic commented:
Maybe put the unittests in the same commits. That'll make it clearer.

david-eckardt-sociomantic committed:
Remove broken and unnecessary static assert

The other static assertion in the function is sufficient. This one causes a
compile error anyway: `S` should be `T`.

david-eckardt-sociomantic committed:
Serializer: Reset dynamic array elements of static array

The array length + pointer ended up in the serialised struct.

david-eckardt-sociomantic commented:
Updated:
- Addressed @gavin-norman-sociomantic's comments.
- Fixed more bugs, most importantly related to static arrays of dynamic arrays.
- Added support for `const(T[][])` and `const(T[])[n]` (which is the same as `const(T[][n])`).
- Fixed and extended `debug(De/SerializationTrace)` printouts.

I still have to look into `Serializer.resetArrayReferences`, which is not covered by the test and looks `const`-suspicious.

gavin-norman-sociomantic commented:
On a very quick look, it seems that this PR mixes bug-fixes and refactorings. Sorry -- I know it'll be hassle -- but these should be separated. We shouldn't merge refactorings into v2.6.x.

nemanja-boric-sociomantic commented:
Maybe it is worth running both D1 and D2 build locally (or on test server where we have RAM) with a profiler and run GetAll on a big channel and see if there's anything that stands out in D2 build. Given that most of the work is done inside the tokyocabinet, maybe it will be obvious what's the difference.

david-eckardt-sociomantic commented:
Fair enough.

david-eckardt-sociomantic commented:
Separating bugfixes and refactorings shouldn't be too difficult.

gavin-norman-sociomantic commented:
Ok great.

david-eckardt-sociomantic committed:
Serializer: Fix counting nested static arrays as dynamic array slices

This caused the data buffer size to be too large.

david-eckardt-sociomantic committed:
Fix undetected const dynamic array in static assert condition

david-eckardt-sociomantic commented:
Updated, reducing the changes to the bug fixes without unrelated refactoring.

david-eckardt-sociomantic committed:
Make unittest structs static

david-eckardt-sociomantic committed:
Contiguous: Fix const(T[]) dynamic arrays not detected

mihails-strasuns-sociomantic commented on src/ocean/util/serialize/contiguous/Serializer.d:
```diff
@@ -622,6 +622,7 @@ struct Serializer
                 {
                     // element is a dynamic array
                     data = This.dumpArray(element, data);
+                    *(cast(U*)&element) = null;
```
Out of curiosity - is there any point in doing so other than simplifying debugging? Deserializer doesn't care if pointers are null or garbage.

mihails-strasuns-sociomantic commented on src/ocean/util/serialize/contiguous/package_.d:
```diff
@@ -696,3 +697,46 @@ unittest
+    ConstS s;
+
+    Serializer.serialize(s, buffer);
+    auto cont_S = Deserializer.deserialize!(ConstS)(buffer);
```
As mentined in other issue, my intention was to not allow this, requiring target struct to be 100% mutable. Exactly because it simplifies (de)serializer code a lot while not being relevant to our typical use cases. What has caused you to pursue such struct support? Original plan was to go with using mutable structs internally and aplying `const` to them as a whole when accepting external params.

david-eckardt-sociomantic replied:
The motivation is to support struct fields containing an array of swarm channels (i.e. strings) where the original channels, passed by the user to the swarm API, can be `immutable(char[])[]`.

david-eckardt-sociomantic replied:
Sorry, I didn't read your comment very well :blush: -- let me check if there was a complication with the original plan.

mihails-strasuns-sociomantic replied:
Please also check with @gavin-norman-sociomantic - we have just had a slack discussion about this and he went to try another approach in dhtproto. It looks like things got a bit of out of control because I expected deserializer to refuse const types and it silently ingored them instead :( Going to see if that alternative approach proves feasible before making decision about serializer.

mihails-strasuns-sociomantic reviewed (commented):
Other than `const` concerns I have mentioned, I would really appreciate if every commit that fixes something included matching regression unittest. Serializer code base is rather fragile and convoluted, it would be risky to live even single line change to chance.

david-eckardt-sociomantic commented:
>  I would really appreciate if every commit that fixes something included matching regression unittest.

A regression test is a test that will fail if a fixed bug is re-added again, correct?

mihails-strasuns-sociomantic commented:
> A regression test is a test that will fail if a fixed bug is re-added again, correct?

Yes. In this context that means that every commit that fixes something should include test that fails without that new fix.

david-eckardt-sociomantic committed:
De/Serializer: Reject and support const(T[]) and const(T[])[n] fields, resp

Handle const(T[]) and const(T[])[n] fields: Support them in the Serializer
and reject them in the Deserializer. Also fix bugs causing compile-time
errors (such as attempting to assign to a const struct field or array
element) and several run-time errors where const(T[]) struct fields or
array elements were not detected to be dynamic arrays and skipped.

david-eckardt-sociomantic commented:
There are still several dynamic array detections which incorrectly do not unqualify the input type so they should detecting `const(T[])` as a dynamic array, but surprisingly they succeed. This is my surprise:
```d
alias const(char[]) T;
static if (is(T U == U[]))
    static assert(is(T == U[])); // fails -- why is it true in the static if?
```
The [D2 specification](http://dlang.org/spec/expression.html#IsExpression) says
> The condition is satisfied if _Type_ [i.e. `T`] is semantically correct and is the same as _TypeSpecialization_ [i.e. `U[]`].

In the example the condition is satisfied but they are not the same as the `static assert` confirms. Is this a compiler bug? I couldn't find an issue on https://issues.dlang.org/ so far.

If it was really a compiler bug it would hide bugs in the `Serializer` code.

david-eckardt-sociomantic commented:
DMD bug reported: https://issues.dlang.org/show_bug.cgi?id=17353

david-eckardt-sociomantic committed:
ContainsDynamicArray: Add missing Unqual in dynamic array detection

This bug is hidden by DMD issue #17353.

david-eckardt-sociomantic commented:
Updated:
- Regression tests
- Added `Unqual` to dynamic array type checks including those that would work by accident due to DMD bug #17353.

mihails-strasuns-sociomantic commented on src/ocean/util/serialize/contiguous/Serializer.d:
```diff
@@ -622,6 +622,7 @@ struct Serializer
                 {
                     // element is a dynamic array
                     data = This.dumpArray(element, data);
+                    *(cast(T*)&element) = null;
```
This is going to be a problem with https://github.com/sociomantic-tsunami/ocean/pull/88 though - or do we have a way to reset all pointers to null in-place?

mihails-strasuns-sociomantic commented:
@david-eckardt-sociomantic I see that it still has changes to support `const` in deserializer - could you please move them to separate PR so that non-controversial serializer fixes can be merged ASAP?

david-eckardt-sociomantic commented:
Updated, now rejecting type qualifiers in `Deserializer`. This changes only one commit, `De/Serializer: Reject and support const(T[]) and const(T[])[n] fields, resp`.

mihails-strasuns-sociomantic commented on src/ocean/util/serialize/contiguous/Deserializer.d:
```diff
@@ -1089,6 +1093,20 @@ struct Deserializer
+    template RejectQualifier ( T )
+    {
+        version (D_Version2)
+            static assert(is(T == typeof(mixin("cast()T.init"))),
```
As this is bug fix release, please make it `pragma` for now!

mihails-strasuns-sociomantic reviewed (changes requested):
One last request :)
